### PR TITLE
osd: Fix the way that auto repair triggers after regular scrub

### PR DESCRIPTION
**Pull request by dzafman**
We used a trick to get auto repair to happen after scrub errors
which reset the scrub/deep-scrub stamps.  This not only
looks bad to the user, but causes health warnings.

Caused by: 2202e5d0b107795837ce79ffce2a980e8c12fc62

Fixes: http://tracker.ceph.com/issues/40073

Signed-off-by: David Zafman <dzafman@redhat.com>

- [x] References tracker ticket
- [x] ~Updates documentation if necessary~
- [x] ~Includes tests for new functionality or reproducer for bug~

### CODE REVIEW

**Comment by dzafman** [comment]
TESTING PASSED:
http://pulpito.ceph.com/dzafman-2019-05-31_07:47:29-rados-wip-zafman-testing-distro-basic-smithi/
valgrind, infrastructure and unrelated issues

**Comment by dzafman** [comment]
jenkins retest make check

**Comment by dzafman** [comment]
jenkins test make check

**Comment by neha-ojha** [comment]
retest this please

File: src/osd/PG.h [hunk 1]
```diff
@@ -1120,7 +1120,7 @@ class PG : public DoutPrefixProvider, public PeeringState::PeeringListener {
     utime_t sleep_start;
 
     // flags to indicate explicitly requested scrubs (by admin)
-    bool must_scrub, must_deep_scrub, must_repair;
+    bool must_scrub, must_deep_scrub, must_repair, need_auto;
```
Comment (tchaikov) [inline src/osd/PG.h]: i feel that `must_scrub` and `need_auto` are mutual-exclusive. can we consolidate them into a single variable? like
```c++
enum class Initiated {
  by_user,
  by_routine,
  on_error,
  none,
};
Initiated why = Initiated::none;
```

actually, the other flags can be populated from this single variable.

**Comment by dzafman** [comment]
PG::sched_scrub() is re-entered as reservations come in, so deep-scrubs due to osd_deep_scrub_randomize_ratio only matters on the final entry.  This is a maintainability issue.  Also, the automatic scrub -> deep_scrub upgrade doesn't include auto_repair handling which I want to fix.

File: src/osd/PG.h (hunk 1 above)
Comment (dzafman) [reply src/osd/PG.h]: @tchaikov need_auto may be superseded by a user request.  So must_scrub could become set in a rare case with a need_auto condition.  The latest code will ignore and then clear need_auto in that rare race case.

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -7155,6 +7155,13 @@ void OSD::sched_scrub()
 	pg->unlock();
 	break;
       }
+      if (pg->is_scrub_registered()) {
+	// Computing backoff based on the normal scrub interval.  By default of 24 hours
+	// we will try again in 10 minutes by dividing by 144.
+	utime_t backoff_time = utime_t(cct->_conf->osd_scrub_min_interval / 144, 0);
```
Comment (dzafman) [inline src/osd/OSD.cc]: Need to check pool min interval too and @neha-ojha suggests using 1/24 for 1 hour by default

**Comment by dzafman** [comment]
jenkins test make check

**Comment by dzafman** [comment]
jenkins test make check

**Comment by dzafman** [comment]
Testing passed:
http://pulpito.ceph.com/dzafman-2019-06-26_07:49:49-rados-wip-zafman-scrub-alt-testing-distro-basic-smithi
All failures are unrelated

**Comment by dzafman** [comment]
jenkins test make check

**Pull request merged**